crates/flux-middle/src/global_env.rs

```diff
@@ -180,63 +183,114 @@ impl<'sess, 'tcx> GlobalEnv<'sess, 'tcx> {
         }
     }
 
-    pub fn sort_of_res(&self, res: fhir::Res) -> Option<fhir::Sort> {
-        // CODESYNC(sort-of, 4) sorts should be given consistently
-        match res {
+    pub fn sort_of_path(&self, path: &fhir::Path) -> Option<fhir::Sort> {
+        // CODESYNC(sort-of-path, 2) sorts should be given consistently
+        match path.res {
             fhir::Res::PrimTy(PrimTy::Int(_) | PrimTy::Uint(_)) => Some(fhir::Sort::Int),
             fhir::Res::PrimTy(PrimTy::Bool) => Some(fhir::Sort::Bool),
             fhir::Res::PrimTy(PrimTy::Float(..) | PrimTy::Str | PrimTy::Char) => {
                 Some(fhir::Sort::Unit)
             }
             fhir::Res::Def(DefKind::TyAlias { .. } | DefKind::Enum | DefKind::Struct, def_id) => {
-                Some(fhir::Sort::Record(def_id))
-            }
-            fhir::Res::SelfTyAlias { alias_to, .. } => {
-                let self_ty = self.tcx.type_of(alias_to).skip_binder();
-                self.sort_of_self_ty(self_ty)
-            }
-            fhir::Res::Def(DefKind::TyParam, def_id) => {
-                let param = self.get_generic_param(def_id.expect_local());
-                match &param.kind {
-                    fhir::GenericParamKind::BaseTy => Some(fhir::Sort::Param(def_id)),
-                    fhir::GenericParamKind::Type { .. } | fhir::GenericParamKind::Lifetime => None,
-                }
+                let mut sort_args = vec![];
+                if let Some(generics) = self.map().get_generics(def_id) {
+                    for (param, arg) in generics.params.iter().zip(&path.args) {
+                        if let GenericParamKind::SplTy = param.kind {
+                            let fhir::GenericArg::Type(ty) = arg else { return None };
+                            let sort = self.sort_of_ty(ty)?;
+                            sort_args.push(sort);
+                        }
+                    }
+                };
+                Some(fhir::Sort::Record(def_id, List::from_vec(sort_args)))
             }
+            fhir::Res::SelfTyAlias { alias_to, .. } => self.sort_of_self_ty_alias(alias_to),
+            fhir::Res::Def(DefKind::TyParam, def_id) => self.sort_of_generic_param(def_id),
             fhir::Res::Def(DefKind::AssocTy | DefKind::OpaqueTy, _)
             | fhir::Res::SelfTyParam { .. } => None,
-            fhir::Res::Def(..) => bug!("unexpected res {res:?}"),
+            fhir::Res::Def(..) => bug!("unexpected res {:?}", path.res),
+        }
+    }
+
+    pub fn sort_of_self_ty_alias(&self, alias_to: DefId) -> Option<fhir::Sort> {
+        let self_ty = self.tcx.type_of(alias_to).instantiate_identity();
+        self.sort_of_self_ty(alias_to, self_ty)
+    }
+
+    fn sort_of_generic_param(&self, def_id: DefId) -> Option<fhir::Sort> {
+        let param = self.get_generic_param(def_id.expect_local());
+        match &param.kind {
+            fhir::GenericParamKind::BaseTy | fhir::GenericParamKind::SplTy => {
```

Comment:
Same comment here about `fhir::Sort::Param` not quite being the way we should represent the sort of a `SplTy`